fvm/evm/emulator/emulator.go

```diff
@@ -132,36 +132,76 @@ func (bl *BlockView) DirectCall(call *types.DirectCall) (*types.Result, error) {
 func (bl *BlockView) RunTransaction(
 	tx *gethTypes.Transaction,
 ) (*types.Result, error) {
-	var res *types.Result
-	var err error
 	proc, err := bl.newProcedure()
 	if err != nil {
 		return nil, err
 	}
-	txHash := tx.Hash()
+
 	msg, err := gethCore.TransactionToMessage(
 		tx,
 		GetSigner(bl.config),
 		proc.config.BlockContext.BaseFee)
 	if err != nil {
 		// this is not a fatal error (e.g. due to bad signature)
 		// not a valid transaction
-		res = &types.Result{
-			TxType: tx.Type(),
-			TxHash: txHash,
-		}
-		res.SetValidationError(err)
-		return res, nil
+		return types.NewInvalidResult(tx, err), nil
 	}
 
 	// update tx context origin
 	proc.evm.TxContext.Origin = msg.From
-	res, err = proc.run(msg, txHash, 0, tx.Type())
+	res, err := proc.run(msg, tx.Hash(), 0, tx.Type())
 	if err != nil {
 		return nil, err
 	}
 	// all commmit errors (StateDB errors) has to be returned
-	return res, proc.commitAndFinalize()
+	if err := proc.commit(true); err != nil {
+		return nil, err
+	}
+
+	return res, nil
+}
+
+func (bl *BlockView) BatchRunTransactions(txs []*gethTypes.Transaction) ([]*types.Result, error) {
+	batchResults := make([]*types.Result, len(txs))
+
+	proc, err := bl.newProcedure()
+	if err != nil {
+		return nil, err
+	}
+
+	for i, tx := range txs {
+		msg, err := gethCore.TransactionToMessage(
+			tx,
+			GetSigner(bl.config),
+			proc.config.BlockContext.BaseFee)
+		if err != nil {
+			batchResults[i] = types.NewInvalidResult(tx, err)
+			continue
+		}
+
+		// update tx context origin
+		proc.evm.TxContext.Origin = msg.From
+		res, err := proc.run(msg, tx.Hash(), uint(i), tx.Type())
+		if err != nil {
+			return nil, err
+		}
+		// all commmit errors (StateDB errors) has to be returned
+		if err := proc.commit(false); err != nil {
+			return nil, err
+		}
+
```

Comment:
I think you have to collect the logs somewhere here.

Reply:
otherwise the events might end up with the wrong tx hash, you might also add this to you the test to check for the tx_hash for events.

Reply:
hmm good idea on adding tests and I added them to compare logs, but I believe logs should be collected since they are being collected as part of the `procedure.run()` https://github.com/onflow/flow-go/blob/80d11a0f97dbed9356572f4fbcc5a9cd174f98a4/fvm/evm/emulator/emulator.go#L490
which happens before reset. It is weird tho that in tests I don't see log values for transaction hash, or block height populated correctly, I'm investigating as to why, but the weird part is when events are being emitted the values are there, so I'm looking into possibility of them being incorrectly decoded in tests or test env, the  test that fails: https://github.com/onflow/flow-go/blob/a20fe793cc3e0cadca493c458a4e23d277926464/fvm/evm/evm_test.go#L371